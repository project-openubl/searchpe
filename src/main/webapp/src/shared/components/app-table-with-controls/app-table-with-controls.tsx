import React, { useMemo } from "react";

import {
  Toolbar,
  ToolbarContent,
  ToolbarGroup,
  ToolbarItem,
  ToolbarItemVariant,
} from "@patternfly/react-core";
import {
  IActions,
  IActionsResolver,
  IAreActionsDisabled,
  ICell,
  IExtraColumnData,
  IRow,
  ISortBy,
  SortByDirection,
} from "@patternfly/react-table";

import { AppTable } from "../app-table/app-table";
import { SimplePagination } from "../simple-pagination";

import { SearchInput } from "./search-input";

export interface AppTableWithControlsProps {
  count: number;
  items: any[];
  itemsToRow: (items: any[]) => IRow[];

  hiddeBottomPagination?: boolean;
  pagination: {
    perPage?: number;
    page?: number;
  };
  sortBy?: ISortBy;
  handleFilterTextChange: (filterText: string) => void;
  handlePaginationChange: ({
    page,
    perPage,
  }: {
    page: number;
    perPage: number;
  }) => void;
  handleSortChange: (
    event: React.MouseEvent,
    index: number,
    direction: SortByDirection,
    extraData: IExtraColumnData
  ) => void;

  columns: ICell[];
  actions?: IActions;
  actionResolver?: IActionsResolver;
  areActionsDisabled?: IAreActionsDisabled;

  isLoading: boolean;
  loadingVariant?: "skeleton" | "spinner" | "none";
  fetchError?: any;

  toolbar?: any;

  filtersApplied: boolean;
  noDataState?: any;
  noSearchResultsState?: any;
  errorState?: any;
}

export const AppTableWithControls: React.FC<AppTableWithControlsProps> = ({
  count,
  items,
  itemsToRow,

  hiddeBottomPagination,
  pagination,
  sortBy,
  handleFilterTextChange,
  handlePaginationChange,
  handleSortChange,

  columns,
  actions,
  actionResolver,
  areActionsDisabled,

  isLoading,
  fetchError,
  loadingVariant,

  toolbar,

  filtersApplied,
  noDataState,
  noSearchResultsState,
  errorState,
}) => {
  const rows = useMemo(() => itemsToRow(items), [items, itemsToRow]);

  return (
    <div style={{ backgroundColor: "var(--pf-global--BackgroundColor--100)" }}>
      <Toolbar>
        <ToolbarContent>
          <ToolbarGroup>
            <ToolbarItem>
              <SearchInput onSearch={handleFilterTextChange} />
            </ToolbarItem>
          </ToolbarGroup>
          {toolbar}
          <ToolbarItem
            variant={ToolbarItemVariant.pagination}
            alignment={{ default: "alignRight" }}
          >
            <SimplePagination
              count={count}
              params={pagination}
              onChange={handlePaginationChange}
              isTop={true}
            />
          </ToolbarItem>
        </ToolbarContent>
      </Toolbar>
      <AppTable
        columns={columns}
        rows={rows}
        actions={actions}
        actionResolver={actionResolver}
        areActionsDisabled={areActionsDisabled}
        isLoading={isLoading}
        fetchError={fetchError}
        loadingVariant={loadingVariant}
        sortBy={sortBy}
        onSort={handleSortChange}
        filtersApplied={filtersApplied}
        noDataState={noDataState}
        noSearchResultsState={noSearchResultsState}
        errorState={errorState}
      />
      {!hiddeBottomPagination && (
        <SimplePagination
          count={count}
          params={pagination}
          onChange={handlePaginationChange}
        />
      )}
    </div>
  );
};
