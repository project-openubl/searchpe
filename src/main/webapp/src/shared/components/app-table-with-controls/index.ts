export { AppTableWithControls } from "./app-table-with-controls";
