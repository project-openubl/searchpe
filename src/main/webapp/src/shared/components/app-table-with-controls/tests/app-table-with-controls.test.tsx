import React from "react";
import { mount, shallow } from "enzyme";
import { ICell } from "@patternfly/react-table";

import { AppTableWithControls } from "../app-table-with-controls";

describe("AppTableWithControls", () => {
  const columns: ICell[] = [{ title: "Col1" }];
  const itemsToRow = (items: string[]) => {
    return items.map((item) => ({
      cells: [
        {
          title: item,
        },
      ],
    }));
  };

  it("Renders without crashing", () => {
    const wrapper = shallow(
      <AppTableWithControls
        count={120}
        items={["first", "second", "third"]}
        itemsToRow={itemsToRow}
        pagination={{ page: 1, perPage: 10 }}
        handleFilterTextChange={jest.fn()}
        handlePaginationChange={jest.fn()}
        handleSortChange={jest.fn()}
        columns={columns}
        isLoading={false}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Renders adding toolbar", () => {
    const wrapper = shallow(
      <AppTableWithControls
        count={120}
        items={["first", "second", "third"]}
        itemsToRow={itemsToRow}
        pagination={{ page: 1, perPage: 10 }}
        handleFilterTextChange={jest.fn()}
        handlePaginationChange={jest.fn()}
        handleSortChange={jest.fn()}
        columns={columns}
        isLoading={false}
        toolbar={<p>This is an additional content to toolbar</p>}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
