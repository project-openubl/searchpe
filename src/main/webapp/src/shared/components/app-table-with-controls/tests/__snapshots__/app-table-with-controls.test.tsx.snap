// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppTableWithControls Renders adding toolbar 1`] = `
<div
  style={
    Object {
      "backgroundColor": "var(--pf-global--BackgroundColor--100)",
    }
  }
>
  <Toolbar>
    <ToolbarContent
      isExpanded={false}
      showClearFiltersButton={false}
    >
      <ForwardRef>
        <ToolbarItem>
          <SearchInput
            onSearch={[MockFunction]}
          />
        </ToolbarItem>
      </ForwardRef>
      <p>
        This is an additional content to toolbar
      </p>
      <ToolbarItem
        alignment={
          Object {
            "default": "alignRight",
          }
        }
        variant="pagination"
      >
        <SimplePagination
          count={120}
          isTop={true}
          onChange={[MockFunction]}
          params={
            Object {
              "page": 1,
              "perPage": 10,
            }
          }
        />
      </ToolbarItem>
    </ToolbarContent>
  </Toolbar>
  <AppTable
    columns={
      Array [
        Object {
          "title": "Col1",
        },
      ]
    }
    isLoading={false}
    onSort={[MockFunction]}
    rows={
      Array [
        Object {
          "cells": Array [
            Object {
              "title": "first",
            },
          ],
        },
        Object {
          "cells": Array [
            Object {
              "title": "second",
            },
          ],
        },
        Object {
          "cells": Array [
            Object {
              "title": "third",
            },
          ],
        },
      ]
    }
  />
  <SimplePagination
    count={120}
    onChange={[MockFunction]}
    params={
      Object {
        "page": 1,
        "perPage": 10,
      }
    }
  />
</div>
`;

exports[`AppTableWithControls Renders without crashing 1`] = `
<div
  style={
    Object {
      "backgroundColor": "var(--pf-global--BackgroundColor--100)",
    }
  }
>
  <Toolbar>
    <ToolbarContent
      isExpanded={false}
      showClearFiltersButton={false}
    >
      <ForwardRef>
        <ToolbarItem>
          <SearchInput
            onSearch={[MockFunction]}
          />
        </ToolbarItem>
      </ForwardRef>
      <ToolbarItem
        alignment={
          Object {
            "default": "alignRight",
          }
        }
        variant="pagination"
      >
        <SimplePagination
          count={120}
          isTop={true}
          onChange={[MockFunction]}
          params={
            Object {
              "page": 1,
              "perPage": 10,
            }
          }
        />
      </ToolbarItem>
    </ToolbarContent>
  </Toolbar>
  <AppTable
    columns={
      Array [
        Object {
          "title": "Col1",
        },
      ]
    }
    isLoading={false}
    onSort={[MockFunction]}
    rows={
      Array [
        Object {
          "cells": Array [
            Object {
              "title": "first",
            },
          ],
        },
        Object {
          "cells": Array [
            Object {
              "title": "second",
            },
          ],
        },
        Object {
          "cells": Array [
            Object {
              "title": "third",
            },
          ],
        },
      ]
    }
  />
  <SimplePagination
    count={120}
    onChange={[MockFunction]}
    params={
      Object {
        "page": 1,
        "perPage": 10,
      }
    }
  />
</div>
`;
